juju/juju
# [JUJU-3715] Clean up worker on delete model

~~Requires https://github.com/juju/juju/pull/15579. Will rebase once landed.~~

----
 
The following defines a DBManager. The DBManager allows for more
management of a db, this currently includes deleting a db. This is
so when we delete a model we actually stop the worker for that model.
No other access will then be allowed on that model as we've closed it.

Originally, the service layer would just remove the model uuid from
the model table on the controller, but that's not actually enough. If
you have 100s of models you'll still have 100s of workers still working
away. This actually cleans up the workers by removing them completely.

To do this in a sound fashion, the state no longer deletes the model
from the model list. Instead, the responsibility for that falls back
to the dbaccessor worker. To ensure that we correctly order the deletion
and then the stopping we use the special powers that the dbaccessor
worker has, with accessing the db directly and ensuring that everything
is serialized in the singular loop. That way we don't end up with
any out-of-order events if the state deletes the model uuid and then
the runner stops the worker. Serialization is key here.

The only probably is now if you've already got access to a tracked
db and someone else closes it whilst you've got a lease to it, you'll
end up with a closed db. This should cause the workers to die. We'll
need to verify that as we move forward.

For now, this is the mechanics for sorting that out.

## Checklist


## QA steps

```sh
$ juju bootstrap lxd test --build-agent
$ juju add-model default
$ juju destroy-model default
```

Ensure everything is cleaned up correctly.
